applications: asset_tracker_v2: Add documentation for the cloud module (#6315)
nrfconnect/sdk-nrf: applications/asset_tracker_v2/doc/cloud_wrapper.rst

.. _api_cloud_wrapper:

Cloud wrapper API
#################

.. contents::
   :local:
   :depth: 2

> The cloud wrapper API is a generic API used to control the connection to a supported |NCS| client library through :ref:`Integration layers <integration_layers>`.
> It exposes generic functions such as ``send``, ``connect``, and ``disconnect`` hiding functionality that is specific to a single client library implementation.

Reviewer: ```suggestion
The cloud wrapper API is a generic API used for controlling the connection to a supported |NCS| client library through :ref:`Integration layers <integration_layers>`.
It exposes generic functions such as ``send``, ``connect``, and ``disconnect``, hiding the functionality that is specific to a single client library implementation.
```

Integration layers
******************

> Table :ref:`Integration layers <integration_layers>` lists the |NCS| client libraries that are supported by the cloud wrapper API togheter with the associated integration layers.

Reviewer: ```suggestion
The :ref:`Integration layers <integration_layers>` table lists the |NCS| client libraries that are supported by the cloud wrapper API and the associated integration layers.
```

> Each integration layer interacts with a specific client library and contains code that is required to properly setup and maintain a connection to the designated cloud service.

Reviewer: ```suggestion
Each integration layer interacts with a specific client library and contains the code required to properly set up and maintain a connection to the designated cloud service.
```

> Based on the Kconfig options listed in :ref:`Integration layers <integration_layers>`, the corresponding integation layer + client library combination is linked in.

Reviewer: ```suggestion
Based on the Kconfig options listed in the table, the corresponding integration layer + client library combination is linked in.
```
Removed the link as you are referring to the table already a couple of lines above.

> For instance, if :kconfig:`CONFIG_AWS_IOT` is enabled CMake links in :file:`asset_tracker_v2/src/cloud/aws_iot_integration.c` that integrates the :ref:`lib_aws_iot` library.

Reviewer: ```suggestion
For example, if :kconfig:`CONFIG_AWS_IOT` is enabled, CMake links in the :file:`asset_tracker_v2/src/cloud/aws_iot_integration.c` file that integrates the :ref:`lib_aws_iot` library.
```

> This happens in :file:`asset_tracker_v2/src/cloud/CMakeLists.txt`.

Reviewer: ```suggestion
This takes place in the :file:`asset_tracker_v2/src/cloud/CMakeLists.txt` file.
```

.. note::

>    The various integration layers all share a common header file, :file:`asset_tracker_v2/src/cloud/cloud_wrapper.h` that exposes generic functions to send and receive data from the integration layer.

Reviewer: ```suggestion
   The various integration layers all share a common header file :file:`asset_tracker_v2/src/cloud/cloud_wrapper.h` that exposes generic functions to send and receive data from the integration layer.
```

>    The various intergration layers all share a common header file, :file:`asset_tracker_v2/src/cloud/cloud_wrapper.h` that exposes generic functions to send and receive data from the integration layer.

Reviewer 2: ```suggestion
   The various integration layers all share a common header file, :file:`asset_tracker_v2/src/cloud/cloud_wrapper.h` that exposes generic functions to send and receive data from the integration layer.
```

+===========================+==================================================================+==================================+
| :ref:`lib_aws_iot`        |   :file:`asset_tracker_v2/src/cloud/aws_iot_integration.c`       | :kconfig:`CONFIG_AWS_IOT`        |
+---------------------------+------------------------------------------------------------------+----------------------------------+
| :ref:`lib_azure_iot_hub`  |   :file:`asset_tracker_v2/src/cloud/azure_iot_hub_integration.c` | :kconfig:`CONFIG_AZURE_IOT_HUB`  |
+---------------------------+------------------------------------------------------------------+----------------------------------+
| :ref:`lib_nrf_cloud`      |   :file:`asset_tracker_v2/src/cloud/nrf_cloud_integration.c`     | :kconfig:`CONFIG_NRF_CLOUD_MQTT` |
+---------------------------+------------------------------------------------------------------+----------------------------------+

Data addressing
***************

Each integration layer routes data to specific endpoints based on the content of the data and the cloud wapper API call.

> The tables in :ref:`Data routing tables <data_filtering>` lists the various endpoints that are used in each cloud service implementation.

Reviewer: ```suggestion
The :ref:`Data routing tables <data_filtering>` list the endpoints that are used in each cloud service implementation.
```

.. note::

>    The *Direction* entry in table :ref:`Data routing tables <data_filtering>` refers to the direction of the data flow.

Reviewer: ```suggestion
   The *Direction* entry in the :ref:`Data routing tables <data_filtering>` refers to the direction of the data flow.
```

+------------------------------+----------------------------------------------------+-----------+
| Buffered sensor/device data  | <imei>/batch                                       | D2C       |
+------------------------------+----------------------------------------------------+-----------+
| A-GPS response               | <imei>/agps                                        | C2D       |
+------------------------------+----------------------------------------------------+-----------+
| P-GPS response               | <imei>/pgps                                        | C2D       |
+------------------------------+----------------------------------------------------+-----------+
| Device configuration updates | $aws/things/<imei>/shadow/delta                    | C2D       |
|                              +----------------------------------------------------+           |
|                              | $aws/things/<imei>/shadow/get/accepted             |           |
|                              +----------------------------------------------------+           |
|                              | $aws/things/<imei>/shadow/get/accepted/desired/cfg |           |

> +------------------------------+----------------------------------------------------+-----------+

Reviewer 2: **suggestion**: to improve readability, split table in two, one for D2C and one for C2D, remove Direction column.

Reviewer: Maybe also have the topics between double backticks. In all tables.

| Sensor/device data           | $iothub/twin/PATCH/properties/reported/ | NA           | D2C       |
+------------------------------+-----------------------------------------+--------------+-----------+
| Device configuration         | $iothub/twin/PATCH/properties/reported/ | NA           | D2C       |
+------------------------------+-----------------------------------------+--------------+-----------+
| Buffered sensor/device data  | devices/<imei>/messages/events/         | batch        | D2C       |
+------------------------------+-----------------------------------------+--------------+-----------+
| A-GPS response               | devices/<imei>/messages/devicebound/    | agps=result  | CD2       |
+------------------------------+-----------------------------------------+--------------+-----------+
| P-GPS response               | devices/<imei>/messages/devicebound/    | pgps=result  | C2D       |
+------------------------------+-----------------------------------------+--------------+-----------+
| Device configuration updates | $iothub/twin/res/<code>/                | NA           | C2D       |

> +------------------------------+-----------------------------------------+--------------+-----------+

Reviewer 2: **suggestion**: to improve readability, split table in two, one for D2C and one for C2D, remove Direction column.

+------------------------------+----------------------------------------------------+-----------+
| Buffered sensor/device data  | <topic_prefix>/<imei>/d2c/batch                    | D2C       |
+------------------------------+----------------------------------------------------+-----------+
| A-GPS response               | <topic_prefix>/<imei>/c2d                          | C2D       |
+------------------------------+----------------------------------------------------+-----------+
| P-GPS response               | <topic_prefix>/<imei>/c2d                          | C2D       |
+------------------------------+----------------------------------------------------+-----------+
| Device configuration updates | $aws/things/<imei>/shadow/delta                    | C2D       |
|                              +----------------------------------------------------+           |
|                              | $aws/things/<imei>/shadow/get/accepted             |           |
|                              +----------------------------------------------------+           |
|                              | $<imei>/shadow/get/accepted/desired/cfg            |           |

> +------------------------------+----------------------------------------------------+-----------+

Reviewer 2: **suggestion**: to improve readability, split table in two, one for D2C and one for C2D, remove Direction column.

+==============================+========================================================+
| A-GPS response               | ``<imei>/agps``                                        |
+------------------------------+--------------------------------------------------------+
| P-GPS response               | ``<imei>/pgps``                                        |
+------------------------------+--------------------------------------------------------+
| Device configuration updates | ``$aws/things/<imei>/shadow/delta``                    |
|                              +--------------------------------------------------------+
|                              | ``$aws/things/<imei>/shadow/get/accepted``             |
|                              +--------------------------------------------------------+
|                              | ``$aws/things/<imei>/shadow/get/accepted/desired/cfg`` |
+------------------------------+--------------------------------------------------------+

Azure IoT Hub topics
====================

For simplicity, the following table omits certain metavalues present in topics and property bags used in Azure IoT Hub.

> For more information regarding MQTT topics and property bags in Azure IoT Hub, refer to the following documentation `Azure IoT Hub MQTT protocol support`_.

Reviewer: ```suggestion
For more information on MQTT topics and property bags in Azure IoT Hub, refer to the `Azure IoT Hub MQTT protocol support`_ documentation.
```

+------------------------------+---------------------------------------------+--------------+
| Buffered sensor/device data  | ``devices/<imei>/messages/events/``         | ``batch``    |
+------------------------------+---------------------------------------------+--------------+

Cloud-to-Device
---------------

+------------------------------+------------------------------------------+----------------+
|               Data           |             Topic                        | Property bag   |
+==============================+==========================================+================+
| A-GPS response               | ``devices/<imei>/messages/devicebound/`` | ``agps=result``|
+------------------------------+------------------------------------------+----------------+
| P-GPS response               | ``devices/<imei>/messages/devicebound/`` | ``pgps=result``|
+------------------------------+------------------------------------------+----------------+
| Device configuration updates | ``$iothub/twin/res/<code>/``             | NA             |
+------------------------------+------------------------------------------+----------------+

nRF Cloud topics
================

> For more information regarding topics utilized in the nRF Cloud connection, refer to the following documentation `nRF Cloud MQTT API`_ for details.

Reviewer: ```suggestion
For more information on topics used in the nRF Cloud connection, refer to the `nRF Cloud MQTT API`_ documentation.
```
